### ISSUE TITLE
[Proposal] Cannot properly write documentation and feature requests simultaneously.

### ISSUE THREAD

**Issue by JayBaileyCS** [neelnanda-io/TransformerLens#238]
### Proposal 

When writing a new feature for TransformerLens, we want to add a demo or add it to the relevant demo as part of the PR. I would like to be able to robustly check the demo works afterwards without running into the "chicken-and-egg problem" described in the next section.

### Motivation

When writing a new feature for TransformerLens, we want to add a demo or add it to the relevant demo as part of the PR. This demo involves installing TransformerLens from Git, so that people can run them on Colabs. However, since in-progress PR's are not currently in TransformerLens, this means you can't test the demos work properly without submitting the PR and hoping for the best. This can be tested in small features by copy-pasting the PR code into Colab and then just assuming the import runs correctly but I don't like the idea of relying on a lack of human error at this step.

### Pitch

I do not currently have a solution in mind for this problem.

### Alternatives

Alternative 1: Rely on individual contributors to write their demos correctly after their code passes automated tests. Assume demos may occasionally break and need fixing as a result.
Alternative 2: Allow features and demos to be made as separate PR's. This solves the robustness issue but increases the risk that contributors, who are volunteers, will forget to write the documentation afterwards or have life get in the way.

### Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)

**Comment by neelnanda-io** [comment]
My recommended solution is to replace the line `pip install
TransformerLens` with `pip install git+
***@***.***_NAME` for testing
purposes - this will install the most up to date version and test with
that.

This doesn't solve the problem that when a PR is merged in, the `pip
install transformer_lens` library is not necessarily updated. I don't have
a great solution to this, plausibly replacing the `pip install
transformer_lens` line in each demo with `pip install git+
https://github.com/neelnanda-io/TransformerLens` is correct, since this
will always install from main and thus keep things in sync? It definitely
seems bad if any time a demo is updated it breaks the demo before the
library version is incremented.

But yeah, it is a mess and I don't have a perfect solution to this, eg
there's no way to automatically test whether you've broken the demo on
merge (we can ask contributors to manually verify it runs from the top when
using `pip install git+THEIR_BRANCH`, and send a screenshot or smth, then
make a PR with the `pip install git+
https://github.com/neelnanda-io/TransformerLens` line maybe).

Another idea would be to do something like have `if
transformer_lens.__version__ >= NEW_VERSION_NUMBER` at the top of the new
demo cells, but that's also ugly

**Comment by JayBaileyCS** [comment]
I've added some info to the README as part of my latest PR explaining this manual verification step, which is at least a decent workaround.

**Comment by ArthurConmy** [comment]
@JayBaileyCS you could develop notebooks that do this:

```
from IPython import get_ipython 
ipython = get_ipython()
try:
    ipython.run_line_magic("pip", "install git+https://github.com/JayBaileyCS/TransformerLens.git@my_new_feature")
    import transformer_lens
except:
    ipython.run_line_magic("pip", "install git+https://github.com/neelnanda-io/TransformerLens.git")
```

Then you could develop the feature like this, and as soon as you merge delete the branch `my_new_feature`. Does that work for you?

**Comment by alan-cooney** [comment]
To note on this - it would be better if most docs are added in the [API docs](https://neelnanda-io.github.io/TransformerLens/transformer_lens.html) (via docstrings), now that I've cleaned it up (waiting on the PRs to be merged currently).

Notebooks are useful, but it's hard for new users to have to go through >10 notebooks to understand all the functionality. Better to have it instantly available with code hints and also in the docs.

Suggest we close for now, and I'll add some more meta-docs (on how to create great documentation for new features).
